### ISSUE TITLE
Flutter 3.0 seems to ignore Dart version constraints

### ISSUE THREAD

**Issue by larssn** [flutter/flutter#103595]
## Steps to Reproduce

Using flutter 3.0 do:

- `flutter create test`
- Set the SDK version (in pubspec.yaml) to: `sdk: ">=2.10.0 <3.0.0"`
- Remove null safety from `main.dart`
- Build

It will crash with:
```
: Error: The parameter 'key' can't have a value of 'null' because of its type 'Key', but the implicit default value is 'null'.
- 'Key' is from 'package:flutter/src/foundation/key.dart' ('../../flutter/packages/flutter/lib/src/foundation/key.dart').
Try adding either an explicit non-'null' default value or the 'required' modifier.
  const MyApp({Key key}) : super(key: key);
                   ^^^
: Error: The parameter 'key' can't have a value of 'null' because of its type 'Key', but the implicit default value is 'null'.
- 'Key' is from 'package:flutter/src/foundation/key.dart' ('../../flutter/packages/flutter/lib/src/foundation/key.dart').
Try adding either an explicit non-'null' default value or the 'required' modifier.
  const MyHomePage({Key key, this.title}) : super(key: key);
                        ^^^
: Error: The parameter 'title' can't have a value of 'null' because of its type 'String', but the implicit default value is 'null'.
Try adding either an explicit non-'null' default value or the 'required' modifier.
  const MyHomePage({Key key, this.title}) : super(key: key);
```

But Dart 2.10 should allow for non-null-safety.

**Expected results:** 
I expect it to build without null safety.

**Actual results:** 
It doesn't.

```
[✓] Flutter (Channel stable, 3.0.0, on Mac OS X 10.13.6 17G14042 darwin-x64, locale en-DK)
    • Flutter version 3.0.0 at /Users/path/workspace/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ee4e09cce0 (2 days ago), 2022-05-09 16:45:18 -0700
    • Engine revision d1b9a6938a
    • Dart version 2.17.0
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
    • Android SDK at /Users/path/Library/Android/sdk
    • Platform android-31, build-tools 32.0.0
    • ANDROID_HOME = /Users/path/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)
    • All Android licenses accepted.

[!] Xcode - develop for iOS and macOS
    ✗ Xcode installation is incomplete; a full installation is necessary for iOS development.
      Download at: https://developer.apple.com/xcode/download/
      Or install Xcode via the App Store.
      Once installed, run:
        sudo xcode-select --switch /Applications/Xcode.app/Contents/Developer
        sudo xcodebuild -runFirstLaunch
    ! CocoaPods 1.9.1 out of date (1.11.0 is recommended).
        CocoaPods is used to retrieve the iOS and macOS platform side's plugin code that responds to your plugin usage on the Dart side.
        Without CocoaPods, plugins will not work on iOS or macOS.
        For more info, see https://flutter.dev/platform-plugins
      To upgrade see https://guides.cocoapods.org/using/getting-started.html#installation for instructions.

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[✓] VS Code (version 1.67.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.0

[✓] Connected device (3 available)
    • SM T510 (mobile) • R52N70RYCGL • android-arm    • Android 11 (API 30)
    • macOS (desktop)  • macos       • darwin-x64     • Mac OS X 10.13.6 17G14042 darwin-x64
    • Chrome (web)     • chrome      • web-javascript • Google Chrome 101.0.4951.54

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

**Comment by darshankawar** [comment]
Thanks for the report. I see same behavior related to null safety.

<details>
<summary> stable, master flutter doctor -v </summary>

```

[✓] Flutter (Channel stable, 3.0.0, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.0.0 at /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ee4e09cce0 (2 days ago), 2022-05-09 16:45:18 -0700
    • Engine revision d1b9a6938a
    • Dart version 2.17.0
    • DevTools version 2.12.2

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[✓] Flutter (Channel master, 3.1.0-0.0.pre.663, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.1.0-0.0.pre.663 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a298bdb417 (64 minutes ago), 2022-05-11 23:14:08 -0400
    • Engine revision 0673b781d2
    • Dart version 2.18.0 (build 2.18.0-106.0.dev)
    • DevTools version 2.13.1
    
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by mraleph** [comment]
If I were to guess there is a bug in how language features which have both `expired: true` and experimental release version set are handled by CFE. NNBD is [one of those](https://github.com/dart-lang/sdk/blob/main/tools/experimental_features.yaml#L153-L162). We probably think that NNBD is now enabled by 2.10 - which is wrong.

**Comment by johnniwinther** [comment]
It sounds likely. The experimental release version was used to opt in the flutter framework at 2.10, but when developed, the expired flag wasn't set.

**Comment by sohrabonline** [comment]
**The experimental release version was used to opt in the flutter framework at 2.10, but when developed, the expired flag wasn't set.**

**Comment by johnniwinther** [comment]
The problem is that the project created with `flutter create test` identifies itself as the package called `test` in its `pubspec.yaml` (first line `name: test`). Since [`package:test`](https://pub.dev/packages/test) was part of the initial migration to null safety, along the `dart:` libraries and `package:flutter`, the experimental release version (`2.10`) is used as the enabling version.

The fix is therefore either to use `2.9` in the `sdk` constraint or to use a different `name` in `pubspec.yaml`.
